### PR TITLE
gh-105587: Remove assertion from `_PyStaticObject_CheckRefcnt`

### PR DESCRIPTION
**Pull request by eduardo-elizondo**
Currently, we are using _PyStaticObject_CheckRefcnt to check the reference count of objects during runtime shutdown. However, this check can be incorrect if an Extension mutates the immortal reference count. Since we can't guarantee that this can't ever happen, let's instead just remove this check. 

* Issue: gh-105587

### CODE REVIEW

**Comment by eduardo-elizondo** [comment]
cc @ericsnowcurrently

**Comment by eduardo-elizondo** [comment]
cc @ericsnowcurrently @kumaraditya303 let me know what you think!

**Comment by ericsnowcurrently** [comment]
Making it a warning is a good idea.

**Comment by kumaraditya303** [comment]
SGTM, I'll review.

File: Include/internal/pycore_global_objects_fini_generated.h [hunk 1]
```diff
@@ -12,9 +12,11 @@ extern "C" {
 static inline void
 _PyStaticObject_CheckRefcnt(PyObject *obj) {
     if (Py_REFCNT(obj) < _Py_IMMORTAL_REFCNT) {
-        _PyObject_ASSERT_FAILED_MSG(obj,
-            "immortal object has less refcnt than expected "
-            "_Py_IMMORTAL_REFCNT");
+        PyObject *repr = PyObject_Repr(obj);
```
Comment (kumaraditya303) [inline Include/internal/pycore_global_objects_fini_generated.h]: You should use `_PyObject_Dump` not repr as earlier code was doing.

Comment (eduardo-elizondo) [reply Include/internal/pycore_global_objects_fini_generated.h]: Great point that looks much better

**Comment by ericsnowcurrently** [comment]
Both of your points are good ones.  I retract my idea. 😄

File: Include/internal/pycore_global_objects_fini_generated.h [hunk 2]
```diff
@@ -12,9 +12,8 @@ extern "C" {
 static inline void
 _PyStaticObject_CheckRefcnt(PyObject *obj) {
     if (Py_REFCNT(obj) < _Py_IMMORTAL_REFCNT) {
-        _PyObject_ASSERT_FAILED_MSG(obj,
-            "immortal object has less refcnt than expected "
-            "_Py_IMMORTAL_REFCNT");
+        fprintf(stderr, "Immortal Object has less refcnt than expected.\n");
+        _PyObject_Dump(obj);
```
Comment (eduardo-elizondo) [inline Include/internal/pycore_global_objects_fini_generated.h:16]: cc @kumaraditya303 done!

**Review by kumaraditya303 — APPROVED** [review]

**Pull request merged**